### DISCUSSION TITLE
Use rete JS framework in programmking tab

### DISCUSSION THREAD

**Discussion by error414** [iNavFlight/inav#9758] [Ideas]
Would be nice to use rete js framework https://retejs.org/ in programming tab, It would be more understable for non programmers who want to use programming framework. 

I would like to start discusion about about it. There are few questions.

1) do we need it, will it help
2) where to save additional information for block, I talk mainly about positions of each block. I would not like save it into eeprom. maybe save extra data to chrome storage in configurator. Or I can imagine to set all blocks as fixed and users will be allowed just create connectyion beetween blocks and change params inside blocks. 
3) use reteJS or other JS framework for blocks https://www.libhunt.com/r/rete
4) or can create extra configurator (new nwjs app) only for programming framework based on reteJS, export from this configurator could be commands for CLI. 

I will try to create proof of concept how it could look like.

**Comment by MrD-RC** [comment]
Honestly, I looked at the example on their website for the chatbot thing. And found it unintuitive. 

I don’t really see a problem with the current system. It has been made to mimic the basic programming within transmitters. So anyone who has made a logical switch on their radio will feel at home with the UI. 

Also, I think using that Rete system could end up looking very cluttered once you have more than a couple of logic conditions.

**Comment by error414** [comment]
The ardublocky is widely use for arduino programming. Node-red is almost same as retejs. Node-red is used for home automation by users who don't know any programming language.

I see around  me that nobody uses logical switches in edgetx, sometimes somebody copy logical switches from youtube. 

Be honest, i have no idea how programming framework will look like in retejs, it's just in my head, I use node-js a lot and I would say it's not as bad. But you are right that programming fw in inav is quite specific. 

I will try create proof of concept of some Pawel's script  how it could look like. As simple as possible.

**Comment by error414** [comment]
I tryed to do proof of conecpet and it was not easy to move logic from programming framework to retejs, there is only one benefit, make connection ("active") beetween logic conditions, everything else had to be same, so you were right. Using retejs is not right way how to improve programming framework

**Reply by sensei-hacker** [reply]
Thanks for testing that!
I think the current interface is unintuitive for most people, but don't have any great ideas on how to improve it.

Unintuitive for non-programmers of course, but also unintuitive for most programmers, who are accustomed to writing lines of code.

**Reply by Scavanger** [reply]
How about a simple scripting language that runs purely in the Configurator and "compiles" the code into the existing logic conditions?
I was thinking of the good old BASIC, of course it wouldn't be a complete language, because e.g. no loops are supported, but you could write something like this, which would be much more intuitive, at least for programmers:

```
IF HOME_DISTANCE > 600 THEN
    SET_VTX_LEVEL 2
ENDIF
```

**Reply by sensei-hacker** [reply]
I was thinking along the same lines.
I would love to see what you come up with.

I don't know if you're familiar with compiler-compilers like yacc, but this is exactly what they are made for - creating new language. You define "when you see this token, do this to create some output".

To make it a better fit for the INAV PLC one might decide that in this language, all IF conditions call a "function".  So it's always of the form:

```
function isRTH:
  # Is flight mode RTH ..

if isRTH:
  do stuff
```

Behind the scenes, the "isRTH" is just an alias for condition 10, or whichever line it ends up on.

**Comment by dz0ny** [comment]
Just idea, but you can start with a CLI code generator hosted on GitHub domain. In theory you just need to write AST translation layer, and everyone would be able to use and modify existing rules in programming tab.
